RFC-003-mapeo-core.md

```diff
@@ -0,0 +1,153 @@
+# 0. About this Document
+
+This document may not accurately reflect the current architecture of 
+Mapeo Core, but it covers the proposed scope of improvements to be
+implemented over the course of 2020.
+
+
+# 1. Problems
+
+Mapeo Core has several major drawbacks:
+
+1. Does not adhere to the Single Responsibility Principle.
+1. Entirely callback-based with no promise support.
+1. Lifecycle management
+1. No compatible RPC.
+
+1.1. SRP
+
+Although there have been improvements, Mapeo Core concerns itself with many
+different aspects of the Mapeo application stack including:
+
+* Importing and Exporting data
+* Observations
+* Media
+* Wifi Sync state
+* Peer state
+* Sync files
+
+This testing surface increases cognitive load for incoming developers, as the
+Mapeo Core package handles all of these without clear separation, except for
+Sync which is handled as a separate internal module. 
+
+Furthermore, critical pieces for sync logic exist as duplicated code in both Mapeo
+Desktop/Mobile client applications, which are necessary for sync to be reliable
+and can cause inconsistent behavior between applications as well as make it
+more difficult to test sync behaviors.
+
+1.2 Callback-based
+
+When Mapeo Core was originally developed, Promises were still quite new in
+Node.js and did not have LTS support. Now, Mapeo Core is stable on Node
+12.x and above, and core Node.js tooling for promises has become more
+developed.
+
+Callbacks require more Node.js knowledge as the approach is quite particular to
+Node.js itself, whereas Promises adopt an easier-to-read and more approachable
+toolset, especially for automated testing and client integration.
+
+1.3 Lifecycle management
+
+Mapeo Core does not manage it's own lifecycle and the lifecycle of components
+it depends upon. Peer required dependencies `osm` and `media` must be the
+correct versions that Mapeo Core respects, which could be incorrect without
+notifying the developer, causing unexpected behavior or crash during program
+execution, when the Mapeo Core library expects a different version of `osm` or
+`media` than is passed to it. This puts a burden on the library consumer to use
+the correct versions of all of these libraries. 
+
+Mapeo Core also does not know about it's own lifecycle or the lifecycle of
+other Mapeo Core instances on a single machine; this is akin to the PostgreSQL
+prompt not being able to manage and create multiple databases. This can cause
+headaches for developers and has resulted in two separate implementations in
+Mapeo Desktop and Mobile of lifecycle management which are in some ways
+reconcilable, but nevertheless rather complex to maintain.
+
+In practice, client applications use Mapeo Server to instantiate a Mapeo Core
+instance, which means that the applications cannot easily control the lifecycle
+of Mapeo Core separately from the server.
+
+1.4. No compatible RPC package
+
+Applications that want to integrate into Mapeo Core need to use some sort of
```

Comment:
```suggestion
Platforms that have separate processes for their renderer and for NodeJS need two-way communication with Mapeo Core (which generally runs on the NodeJS process). Right now HTTP is used via Mapeo Server, but a simpler RPC-like interface could be used to properly retrieve and control the Mapeo Core NodeJS process.
```

Reply:
What I wrote here is maybe awkward, but I was trying to

1. outline the underlying constraint
2. identify our current solution
3. identify why our current solution isn't very good, before finally..
4. identify the proposed solution

Reply:
RPC is already used in Desktop and Mobile, they're just implemented separately T_T